spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapper.java

```diff
@@ -0,0 +1,83 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.integration;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.HeaderMapper;
+import org.springframework.util.Assert;
+
+public class PubSubHeaderMapper implements HeaderMapper<Map<String, String>> {
+
+	/**
+	 * Headers to exclude in the {@link #fromHeaders(MessageHeaders, Map)}.
+	 */
+	private String[] doNotMapHeaderNames = {
+			MessageHeaders.ID, MessageHeaders.TIMESTAMP
+	};
+
+	private boolean filterHeaders;
+
+	/**
+	 * Set the names of the headers to filtered out in {@link #fromHeaders(MessageHeaders, Map)}.
+	 * @param doNotMapHeaderNames
+	 */
+	public void setDoNotMapHeaderNames(String... doNotMapHeaderNames) {
+		Assert.notNull(doNotMapHeaderNames, "Header names can't be null.");
+		Assert.noNullElements(doNotMapHeaderNames, "No header name can be null.");
+		this.doNotMapHeaderNames = doNotMapHeaderNames;
+	}
+
+	/**
+	 * Generate headers in {@link com.google.pubsub.v1.PubsubMessage} format from
+	 * {@link MessageHeaders}. All headers are converted into strings.
+	 *
+	 * <p>May filter out designated headers, depending on whether {@code filterHeaders} is
+	 * {@code true}.
+	 * @param messageHeaders headers to map from
+	 * @param pubsubMessageHeaders headers in their final format
+	 */
+	@Override
+	public void fromHeaders(MessageHeaders messageHeaders,
+			final Map<String, String> pubsubMessageHeaders) {
+		messageHeaders.forEach((key, value) -> pubsubMessageHeaders.put(key, value.toString()));
+
+		if (this.filterHeaders) {
+			Arrays.stream(this.doNotMapHeaderNames).forEach(messageHeaders::remove);
```

Comment:
What is the goal to remove headers from the original message?
I thought an idea was like do not map headers to the target `pubsubMessageHeaders`...

Must be `messageHeaders.stream().filter().forEach()`.

No? What am I missing?

And why do we need this `filterHeaders` at all?

Reply:
Oops, my mistake, the code was intended to remove headers from `pubsubMessageHeaders`.

`filterHeaders` is needed to disable header filtering. In fact, the default for the outbound channel adapter will be `false`.

Reply:
So the first `forEach()`, besides copying all the headers, also casts the `Map<String, Object>` to `Map<String, String>`.

I like having the filtering out because otherwise, we'd have to do O(n) passes over `doNotMapHeaderNames`, since it isn't a set but an array.

I prefer to leave things like this, let me know if you see any issues though.

Reply:
That's not what we do in other implementations.
Actually we rely on the `pattern`. And if you want to map all, you use `*`.

The mentioned `StompHeaderMapper` should be a good sample.

Right, you can use `*` by default.

Reply:
Sorry, I'm not following.
What do you mean with `pattern` and `*`?
After looking at that Stomp header mapper, I'm still not sure what you mean or what the issue with the current code is.
Thanks!

Reply:
The issue with the current solution that it doesn't follow a pattern we have in other implementations.

It definitely doesn't matter if we map or filter. From the math perspective it's really the same, but from consistency it would be better to follow what we have so far in other places. 

We do expose `headerPatternsToMap` and allow to configure the patters, for example `foo_*`. Therefore all the headers with the `foo_` prefix will be mapped. 

Your current solution requires from me to map all my `foo_` header names.

You are missing the fact of pattern matching in the mentioned `StompHeaderMapper`:
https://github.com/spring-projects/spring-integration/blob/master/spring-integration-stomp/src/main/java/org/springframework/integration/stomp/support/StompHeaderMapper.java#L212

If you still would like to `filter`, we suggest a solution with the `!` operator: https://docs.spring.io/spring-integration/docs/5.0.4.RELEASE/reference/html/amqp.html#amqp-message-headers